### PR TITLE
light: minor fixes / standardising errors

### PR DESCRIPTION
**Pull request by cmwaters**
## Description

I'm just doing a self audit of the light client. There's a few things I've changed

- Validate trust level in `VerifyNonAdjacent` function
- Make errNoWitnesses public (it's something people running software on top of a light client should be able to parse)
- Remove `ChainID` check of witnesses on start up. We do this already when we compare the first header with witnesses
- Remove `ChainID()` from provider interface

Closes: #4538

### CODE REVIEW

**Comment by cmwaters** [comment]
@melekes we don't really have any reason to have `ChainID()` in the provider interface. We check the chain ID for every block in any case. Perhaps we should remove it or replace it with some other identifier that we can use when returning errors

**Comment by melekes** [comment]
> @melekes we don't really have any reason to have ChainID() in the provider interface

agree. we can remove it

File: light/client.go [hunk 1]
```diff
@@ -977,7 +977,7 @@ func (c *Client) lightBlockFromPrimary(ctx context.Context, height int64) (*type
 	l, err := c.primary.LightBlock(ctx, height)
 	c.providerMutex.Unlock()
 	if err != nil {
-		c.logger.Debug("Error on light block request from primary", "error", err)
+		c.logger.Debug("Error on light block request from primary", "error", err, "provider", c.primary)
```
Comment (melekes) [inline light/client.go:980]:
```suggestion
		c.logger.Debug("Error on light block request from primary", "error", err, "primary", c.primary)
```

File: light/provider/provider.go [hunk 2]
```diff
@@ -9,8 +9,9 @@ import (
 // Provider provides information for the light client to sync (verification
 // happens in the client).
 type Provider interface {
-	// ChainID returns the blockchain ID.
-	ChainID() string
+	// String returns an identifier for the provider. This should be unique.
+	// This is only used for logging and returning errors.
+	String() string
```
Comment (melekes) [inline light/provider/provider.go:14]: this probably should not be required by the interface

Comment (melekes) [reply light/provider/provider.go:14]: if it's only used for logging then it's fine to not require it from implementations and simply do `fmt.Sprintf("%v", provider)` which will call `String()` if such is defined by the struct.

Comment (cmwaters) [reply light/provider/provider.go:14]: ok yeah I was unsure whether we should enforce that the provider also adhere to the stringer interface. I will remove it

**Review by melekes — APPROVED** [review]
👍

File: light/store/db/db.go [hunk 3]
```diff
@@ -238,7 +238,7 @@ func (s *dbs) Prune(size uint16) error {
 		append(s.lbKey(1<<63-1), byte(0x00)),
 	)
 	if err != nil {
-		return err
+		panic(err)
```
Comment (melekes) [inline light/store/db/db.go:241]: How’s panic better than returning ?

Comment (cmwaters) [reply light/store/db/db.go:241]: It's just more consistent with how we handle iterator errors in other methods

**Comment by melekes** [comment]
```
--- FAIL: TestPEXReactorDoesNotDisconnectFromPersistentPeerInSeedMode (0.02s)
    testing.go:915: TempDir RemoveAll cleanup: unlinkat /tmp/TestPEXReactorDoesNotDisconnectFromPersistentPeerInSeedMode672584460/002: directory not empty
FAIL
coverage: 78.7% of statements
FAIL	github.com/tendermint/tendermint/p2p/pex	15.710s
```

unrelated. @cmwaters can you fix this?

**Comment by cmwaters** [comment]
> ```
> --- FAIL: TestPEXReactorDoesNotDisconnectFromPersistentPeerInSeedMode (0.02s)
>     testing.go:915: TempDir RemoveAll cleanup: unlinkat /tmp/TestPEXReactorDoesNotDisconnectFromPersistentPeerInSeedMode672584460/002: directory not empty
> FAIL
> coverage: 78.7% of statements
> FAIL	github.com/tendermint/tendermint/p2p/pex	15.710s
> ```
> 
> unrelated. @cmwaters can you fix this?

Sure - I guess this comes from alessios PR

File: light/errors.go [hunk 4]
```diff
@@ -44,7 +44,7 @@ func (e ErrInvalidHeader) Error() string {
 // with any of the connected witnesses.
 var ErrFailedHeaderCrossReferencing = errors.New("all witnesses have either not responded, don't have the " +
 	" blocks or sent invalid blocks. You should look to change your witnesses" +
-	"  or review the light client's logs for more information")
+	" or review the light client's logs for more information")
```
Comment (marbar3778) [inline light/errors.go:47]: nit: can this be a multi line string?

File: light/store/db/db.go (hunk 3 above)
Comment (marbar3778) [reply light/store/db/db.go:241]: While more consistent do we want to panic on this error? Is it something that would cause a node failure?

**Comment by cmwaters** [comment]
> This pull request **fixes 1 alert** when merging [f0ae59e](https://github.com/tendermint/tendermint/commit/f0ae59e8759ad22ed0b2270c5cf4c0d27ca6bbda) into [b1bbd37](https://github.com/tendermint/tendermint/commit/b1bbd375191bf433ecf340948b5c5ef98bc552f2) - [view on LGTM.com](https://lgtm.com/projects/g/tendermint/tendermint/rev/pr-667c3a38e7d1f261d22360f25b817d39bfcb385e)
> 
> **fixed alerts:**
> 
> * 1 for Clear-text logging of sensitive information

Cheers bro. I think I understood that the first 10 times you said it

**Pull request merged**